mio/src/event_loop.rs

```diff
@@ -0,0 +1,342 @@
+
+static NEXT_LOOP_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+scoped_thread_local!(static CURRENT_LOOP: Loop);
+
+const SLAB_CAPACITY: usize = 1024 * 64;
+
+pub struct Loop {
+    id: usize,
+    active: Cell<bool>,
+    io: RefCell<mio::Poll>,
+    tx: mio::channel::Sender<Message>,
+    rx: mio::channel::Receiver<Message>,
+    dispatch: RefCell<Slab<Scheduled, usize>>,
+}
+
+#[derive(Clone)]
+pub struct LoopHandle {
+    id: usize,
+    tx: mio::channel::Sender<Message>,
+}
+
+#[derive(Copy, Clone)]
+pub enum Direction {
+    Read,
+    Write,
+}
+
+struct Scheduled {
+    source: Source,
+    reader: Option<Waiter>,
+    writer: Option<Waiter>,
+}
+
+impl Scheduled {
+    fn waiter_for(&mut self, dir: Direction) -> &mut Option<Waiter> {
+        match dir {
+            Direction::Read => &mut self.reader,
+            Direction::Write => &mut self.writer,
+        }
+    }
+
+    fn event_set(&self) -> mio::EventSet {
+        let mut set = mio::EventSet::none();
+        if self.reader.is_some() {
+            set = set | mio::EventSet::readable()
+        }
+        if self.writer.is_some() {
+            set = set | mio::EventSet::writable()
+        }
+        set
+    }
+}
+
+enum Message {
+    AddSource(Source, Arc<AtomicUsize>, Waiter),
+    DropSource(usize),
+    Schedule(usize, Direction, Waiter),
+    Deschedule(usize, Direction),
+    Shutdown,
+}
+
+fn register(poll: &mut mio::Poll, token: usize, sched: &Scheduled) {
+    // TODO: handle error
+    poll.register(&*sched.source,
+                  mio::Token(token),
+                  mio::EventSet::none(),
+                  mio::PollOpt::level())
+        .unwrap();
+}
+
+fn reregister(poll: &mut mio::Poll, token: usize, sched: &Scheduled) {
+    // TODO: handle error
+    poll.reregister(&*sched.source,
+                    mio::Token(token),
+                    sched.event_set(),
+                    mio::PollOpt::edge() | mio::PollOpt::oneshot())
+        .unwrap();
+}
+
+fn deregister(poll: &mut mio::Poll, sched: &Scheduled) {
+    // TODO: handle error
+    poll.deregister(&*sched.source).unwrap();
+}
+
+impl Loop {
+    pub fn new() -> io::Result<Loop> {
+        let (tx, rx) = mio::channel::from_std_channel(mpsc::channel());
+        let io = try!(mio::Poll::new());
+        try!(io.register(&rx,
+                         mio::Token(0),
+                         mio::EventSet::readable(),
+                         mio::PollOpt::edge()));
+        Ok(Loop {
+            id: NEXT_LOOP_ID.fetch_add(1, Ordering::Relaxed),
+            active: Cell::new(true),
+            io: RefCell::new(io),
+            tx: tx,
+            rx: rx,
+            dispatch: RefCell::new(Slab::new_starting_at(1, SLAB_CAPACITY)),
+        })
+    }
+
+    pub fn handle(&self) -> LoopHandle {
+        LoopHandle {
+            id: self.id,
+            tx: self.tx.clone(),
+        }
+    }
+
+    pub fn run<F: Future>(self, f: F) -> Result<F::Item, F::Error> {
```

Comment:
That's what I had at first, but I got uneasy about the leftover state persisting across invocations. (Shutdown can happen with arbitrary stuff potentially in flight). Maybe it's ok, though!